File: DFS&BFS부수기/BOJ_1039.java
package BOJ_1039;

import java.io.*;
import java.util.ArrayDeque;
import java.util.StringTokenizer;

public class BOJ_1039 {

    // https://www.acmicpc.net/problem/1039
    // input
    private static BufferedReader br;

    // variables
    private static int N, K, ans, Nlen;
    private static String Nstr;
    private static final int MAX = 1_000_001;

    private static class Num {
        String num;
        int count;

        private Num(String num, int count) {
            this.num = num;
            this.count = count;
        }
    } // End of Num class

    public static void main(String[] args) throws IOException {
        System.setIn(new FileInputStream("C:\\Users\\bigyo\\Desktop\\알고리즘\\JavaAlgorithm\\src\\BOJ_1039\\res.txt"));
        br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        input();

        bw.write(solve());
        bw.close();
    } // End of main()

    private static String solve() {
        StringBuilder sb = new StringBuilder();

        BFS();
        sb.append(ans);
        return sb.toString();
    } // End of solve()

    private static void BFS() {
        ArrayDeque<Num> que = new ArrayDeque<>();
        boolean[][] isVisited = new boolean[MAX][K];
        que.offer(new Num(Nstr, 0));
        isVisited[N][0] = true;

        while (!que.isEmpty()) {
            Num cur = que.poll();

            for (int i = 0; i < Nlen - 1; i++) {
                for (int j = i + 1; j < Nlen; j++) {
                    if (i == 0 && cur.num.charAt(j) == '0') continue;
                    String swapStr = swap(cur.num, i, j);
                    int swapInt = Integer.parseInt(swapStr);

                    if(cur.count + 1 == K) {
                        ans = Math.max(ans, swapInt);
                        continue;
                    }

                    if (isVisited[swapInt][cur.count + 1]) continue;
                    isVisited[swapInt][cur.count + 1] = true;
                    que.offer(new Num(swapStr, cur.count + 1));
                }
            }
        }
    } // End of BFS()

    private static String swap(String str, int i, int j) {
        char[] chArr = str.toCharArray();
        char ch = chArr[i];
        chArr[i] = chArr[j];
        chArr[j] = ch;
        return new String(chArr);
    } // End of swap()

    private static void input() throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());
        ans = -1;

        Nstr = Integer.toString(N);
        Nlen = Nstr.length();
    } // End of input()
} // End of Main class

File: DFS&BFS부수기/BOJ 1039.kt
package BOJ_1039

import java.io.File
import java.util.StringTokenizer

// https://www.acmicpc.net/problem/1039
// input
private var br = System.`in`.bufferedReader()

// variables
private var N = 0
private var K = 0
private lateinit var chArr: CharArray
private const val MAX = 1_000_001
private data class Swap(val num: String, val count: Int)

fun main() {
    val path = "C:\\Users\\bigyo\\Desktop\\알고리즘\\KotlinAlgo\\src\\main\\kotlin\\BOJ_1039\\res.txt"
    br = File(path).bufferedReader()
    val bw = System.out.bufferedWriter()

    input()

    bw.write(solve())
    bw.close()
} // End of main()

private fun solve(): String {
    val sb = StringBuilder()

    sb.append(BFS())
    return sb.toString()
} // End of solve()

private fun BFS(): Int {
    val que = ArrayDeque<Swap>()
    val memo = Array(MAX) { BooleanArray(K + 1) }
    var ans = -1

    que.addLast(Swap(String(chArr), 0))


    while (que.isNotEmpty()) {
        val cur = que.removeFirst()

        if (cur.count == K) {
            ans = Math.max(ans, cur.num.toInt())
        }
        if (memo[cur.num.toInt()][cur.count]) continue
        memo[cur.num.toInt()][cur.count] = true

        for (i in 0 until N - 1) {
            for (j in i + 1 until N) {
                if (i == 0 && cur.num[j] == '0') continue
                if (cur.count + 1 > K) continue

                val swapNum = swap(cur.num, i, j)
                que.addLast(Swap(swapNum, cur.count + 1))
            }
        }
    }

    return ans
} // End of BFS()

private fun swap(str: String, i: Int, j: Int): String {
    val arr = str.toCharArray()
    val temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp

    return String(arr)
} // End of swap()

private fun input() {
    StringTokenizer(br.readLine()).run {
        chArr = nextToken().toCharArray()
        N = chArr.size
        K = nextToken().toInt()
    }
} // End of input()
